contracts/governance/test/ZeroExVotesMigration.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+
+  Copyright 2023 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+pragma solidity ^0.8.19;
+
+import {ZeroExVotes} from "../src/ZeroExVotes.sol";
+import {SafeCast} from "@openzeppelin/utils/math/SafeCast.sol";
+import {Math} from "@openzeppelin/utils/math/Math.sol";
+import {CubeRoot} from "./CubeRoot.sol";
+
+contract ZeroExVotesMigration is ZeroExVotes {
+    uint32 public migrationBlock;
+
+    function initialize(address, uint256) public virtual override {
+        revert();
+    }
+    function initialize() public virtual onlyProxy reinitializer(2) {
+        migrationBlock = uint32(block.number);
+    }
+
+    struct CheckpointMigration {
+        uint32 fromBlock;
+        uint96 votes;
+        uint96 quadraticVotes;
+        uint32 migratedVotes;
+    }
+
+    function _toMigration(Checkpoint storage ckpt) internal pure returns (CheckpointMigration storage result) {
+        assembly {
+            result.slot := ckpt.slot
+        }
+    }
+    function _toMigration(Checkpoint[] storage ckpt) internal pure returns (CheckpointMigration[] storage result) {
+        assembly {
+            result.slot := ckpt.slot
+        }
+    }
+
+    function getMigratedVotes(address account) public view returns (uint256) {
+        uint256 pos = _checkpoints[account].length;
+        if (pos == 0) {
+            return 0;
+        }
+        Checkpoint storage ckpt = _unsafeAccess(_checkpoints[account], pos - 1);
+        if (ckpt.fromBlock <= migrationBlock) {
+            return 0;
+        }
+        return _toMigration(ckpt).migratedVotes;
+    }
+
+    function getPastMigratedVotes(address account, uint256 blockNumber) public view returns (uint256) {
+        require(blockNumber < block.number, "ZeroExVotesMigration: block not yet mined");
+        if (blockNumber <= migrationBlock) {
+            return 0;
+        }
+
+        (bool success, Checkpoint storage checkpoint) = _checkpointsLookup(_checkpoints[account], blockNumber);
+        if (!success) {
+            return 0;
+        }
+        if (checkpoint.fromBlock <= migrationBlock) {
+            return 0;
+        }
+        return _toMigration(checkpoint).migratedVotes;
+    }
+
+    // TODO: we're not handling totalSupply
+
+    // TODO: need to return the migrated weight
+    function _writeCheckpoint(
```

Comment:
This function is almost entirely copy-pasted from `ZeroExVotes`, except for the bits handling `.migratedVotes`